File: frontenddevops/src/app/modules/cours/services/cours.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';
import { Cours, PageCours } from '../models/cours';

@Injectable({
  providedIn: 'root'
})
export class CoursService {
  private apiUrl = environment.apiUrls.microservicecour + 'cours'

  constructor(private http: HttpClient) { }

  // Créer un nouveau cours
  createCours(cours: Cours): Observable<Cours> {
    return this.http.post<Cours>(this.apiUrl, cours);
  }

  // Mettre à jour un cours
  updateCours(cours: Cours): Observable<Cours> {
    return this.http.put<Cours>(`${this.apiUrl}/${cours.id}`, cours);
  }

  // Supprimer un cours
  deleteCours(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Obtenir un cours par ID
  getCoursById(id: number): Observable<Cours> {
    return this.http.get<Cours>(`${this.apiUrl}/${id}`);
  }

  // Obtenir tous les cours
  getAllCours(): Observable<Cours[]> {
    return this.http.get<Cours[]>(this.apiUrl);
  }

  // Obtenir les cours avec pagination
  getAllCoursPagines(page: number, size: number): Observable<any> {
    return this.http.get(`${this.apiUrl}?page=${page}&size=${size}`);
  }

  // Obtenir les cours par professeur
  getCoursByProfesseur(professeurId: number): Observable<Cours[]> {
    return this.http.get<Cours[]>(`${this.apiUrl}/professeur/${professeurId}`);
  }

  // Obtenir les cours par classe
  getCoursByClasse(classeId: number): Observable<Cours[]> {
    return this.http.get<Cours[]>(`${this.apiUrl}/classe/${classeId}`);
  }

  // Obtenir un cours par code
  getCoursByCode(code: string): Observable<Cours> {
    return this.http.get<Cours>(`${this.apiUrl}/code/${code}`);
  }
}

File: microservice-gestion-cours/src/main/java/com/isi/service/CoursService.java
package com.isi.service;

import com.isi.dto.CoursDto;
import com.isi.model.Cours;
import org.springframework.data.domain.Page;

import java.util.List;
import java.util.Optional;

public interface CoursService {
    Cours saveCours(CoursDto cours);
    Cours updateCours(Long id, CoursDto coursDTO);
    void deleteCours(Long id);
    Cours getCoursById(Long id);
    List<Cours> getAllCours();
    Page<Cours> getAllCoursPagines(int page, int size);
    List<Cours> getCoursByProfesseur(Long professeurId);
    List<Cours> getCoursByClasse(Long classeId);
    Optional<Cours> findByCode(String code);
    boolean existsByCode(String code);
}
